File: backend/app/compliance/rules/uniform_rule.py
from __future__ import annotations

from app.compliance.compliance_rules import COMPLIANCE_RULE_IDS
from app.compliance.config import get_compliance_settings
from app.compliance.rules.base import BaseComplianceRule
from app.compliance.types import ComplianceContext, ComplianceEvaluationResult
from app.compliance.uniform_matcher import match_uniform
from app.models import CameraZone, UniformTemplate


class UniformRule(BaseComplianceRule):
    id = COMPLIANCE_RULE_IDS["UNIFORM_VIOLATION"]
    name = "Sai đồng phục bảo hộ"
    event_type = "UNIFORM_VIOLATION"

    def evaluate(self, context: ComplianceContext) -> ComplianceEvaluationResult:
        if context.metadata.get("trigger_event") != "PERSON_ENTER":
            return ComplianceEvaluationResult(violated=False, score=0.0, evidence={})

        zone = context.db.get(CameraZone, context.zone_id)
        if not zone or not zone.required_uniform_id:
            return ComplianceEvaluationResult(
                violated=False,
                score=1.0,
                evidence={"reason": "no_required_uniform"},
            )

        template = context.db.get(UniformTemplate, zone.required_uniform_id)
        if not template:
            return ComplianceEvaluationResult(
                violated=False,
                score=0.0,
                evidence={"reason": "template_not_found", "required_uniform_id": zone.required_uniform_id},
            )

        person_image = context.metadata.get("person_snapshot")
        if isinstance(person_image, str):
            person_image = person_image.encode("utf-8")

        settings = get_compliance_settings()
        match = match_uniform(
            person_image,
            template.image_paths or [],
            track_id=context.track_id,
            template_id=template.id,
            threshold=settings.uniform_threshold,
        )

        if match.score >= settings.uniform_threshold:
            return ComplianceEvaluationResult(
                violated=False,
                score=match.score,
                evidence={
                    "required_uniform_id": template.id,
                    "required_uniform_name": template.name,
                    "matched": match.matched,
                },
            )

        return ComplianceEvaluationResult(
            violated=True,
            score=match.score,
            evidence={
                "event_type": self.event_type,
                "required_uniform_id": template.id,
                "required_uniform_name": template.name,
                "matched": match.matched,
                "threshold": settings.uniform_threshold,
            },
        )

File: backend/src/modules/compliance/rules/uniformRule.js
import { BaseRule, COMPLIANCE_RULE_IDS } from '../BaseRule.js'
import { getComplianceConfig } from '../../../config/compliance.js'
import { matchUniform } from '../uniformMatcher.js'

export class UniformRule extends BaseRule {
  constructor() {
    super({
      id: COMPLIANCE_RULE_IDS.UNIFORM_VIOLATION,
      name: 'Sai đồng phục bảo hộ',
      eventType: 'UNIFORM_VIOLATION',
    })
  }

  async evaluate(context) {
    if (context.triggerEvent !== 'PERSON_ENTER') {
      return { violated: false, score: 0, evidence: {} }
    }

    const requiredUniformId = context.requiredUniformId
    if (!requiredUniformId) {
      return { violated: false, score: 1, evidence: { reason: 'no_required_uniform' } }
    }

    const templateImages = context.templateImages || []
    const match = await matchUniform(context.personSnapshot || null, templateImages, {
      trackId: context.trackId,
      templateId: requiredUniformId,
    })

    const config = getComplianceConfig()
    if (match.score >= config.uniformThreshold) {
      return { violated: false, score: match.score, evidence: { matched: match.matched } }
    }

    return {
      violated: true,
      score: match.score,
      evidence: { eventType: 'UNIFORM_VIOLATION', matched: match.matched },
    }
  }
}
